core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -1755,6 +1758,20 @@ private static boolean sameRelCollectionNoOrderConsidered(
     return true;
   }
 
+  private static @Nullable List<RexNode> compenProjs(List<RexNode> queryRexNodes,
+      List<RexNode> targetRexNodes) {
+    List<RexNode> compenProjs = new ArrayList<>();
+    for (int i = 0; i < queryRexNodes.size(); i++) {
+      RexNode rexNode = queryRexNodes.get(i);
+      int index = targetRexNodes.indexOf(rexNode);
+      if (index < 0) {
+        return null;
+      }
+      compenProjs.add(new RexInputRef(index, rexNode.getType()));
+    }
+    return compenProjs;
+  }
+
```

Comment:
I had the impression that it was the responsibility of the compensation shuttle (`getRexShuttle`) to do this work. Why is it necessary to have a separate method?